core/iwasm/libraries/debug-engine/debug_engine.c

```diff
@@ -0,0 +1,1011 @@
+    os_mutex_unlock(&control_thread->wait_lock);
+    os_cond_signal(&control_thread->wait_cond);
+    os_thread_join(control_thread->tid, NULL);
+    wasm_runtime_free(control_thread->server);
+
+    os_mutex_destroy(&control_thread->wait_lock);
+    os_cond_destroy(&control_thread->wait_cond);
+    wasm_runtime_free(control_thread);
+}
+
+static WASMDebugEngine *g_debug_engine;
+
+static WASMDebugEngine *
+wasm_debug_engine_create()
+{
+    WASMDebugEngine *engine;
+
+    if (!(engine = wasm_runtime_malloc(sizeof(WASMDebugEngine)))) {
+        LOG_ERROR("WASM Debug Engine error: failed to allocate memory");
+        return NULL;
+    }
+    memset(engine, 0, sizeof(WASMDebugEngine));
+
+    /* TODO: support Wasm platform in LLDB */
+    /*
+    engine->control_thread =
+        wasm_debug_control_thread_create((WASMDebugObject *)engine);
+    engine->control_thread->debug_engine = (WASMDebugObject *)engine;
+    engine->control_thread->debug_instance = NULL;
+    sprintf(engine->control_thread->ip_addr, "127.0.0.1");
+    engine->control_thread->port = 1234;
+    */
+
+    bh_list_init(&engine->debug_instance_list);
+    return engine;
+}
+
+bool
+wasm_debug_engine_init(char *ip_addr, int platform_port, int process_port)
+{
+    if (g_debug_engine == NULL)
+        g_debug_engine = wasm_debug_engine_create();
+
+    if (g_debug_engine) {
+        process_port -= 1;
+        g_debug_engine->platform_port =
+          platform_port > 0 ? platform_port : 1234;
+        g_debug_engine->process_base_port =
+          process_port > 0 ? process_port : 6169;
+        if (ip_addr)
+            sprintf(g_debug_engine->ip_addr, "%s", ip_addr);
+        else
+            sprintf(g_debug_engine->ip_addr, "%s", "127.0.0.1");
+        g_debug_engine->active = true;
+    }
+
+    return g_debug_engine != NULL ? true : false;
+}
+
+void
+wasm_debug_set_engine_active(bool active)
+{
+    if (g_debug_engine) {
+        g_debug_engine->active = active;
+    }
+}
+
+bool
+wasm_debug_get_engine_active(void)
+{
+    if (g_debug_engine) {
+        return g_debug_engine->active;
+    }
+    return false;
+}
+
+void
+wasm_debug_engine_destroy()
+{
+    if (g_debug_engine) {
+        wasm_runtime_free(g_debug_engine);
+        g_debug_engine = NULL;
+    }
+}
+
+/* A debug Instance is a debug "process" in gdb remote protocol
+   and bound to a runtime cluster */
+WASMDebugInstance *
+wasm_debug_instance_create(WASMCluster *cluster)
+{
+    WASMDebugInstance *instance;
+    WASMExecEnv *exec_env;
+
+    if (!g_debug_engine || !g_debug_engine->active) {
+        return NULL;
+    }
+
+    if (!(instance = wasm_runtime_malloc(sizeof(WASMDebugInstance)))) {
+        LOG_ERROR("WASM Debug Engine error: failed to allocate memory");
+        return NULL;
+    }
+    memset(instance, 0, sizeof(WASMDebugInstance));
+    instance->cluster = cluster;
+    instance->control_thread =
+      wasm_debug_control_thread_create((WASMDebugObject *)instance);
```

Comment:
the idea of forced conversion between the `WASMDebugObject` and `WASMDebugInstance` is a huge red flag to me.  It's confusing, error-prone and in fact doesn't work. e.g. there's no guarantee that the compiler will generate exactly same layout for the two different structs. Having identical member in identical order doesn't automatically get that guarantee, although it *seems* work for most compilers. I just don't understand why ding this instead of simply aggregate them. btw, there's no document for the relationship between the two structs either. IMO it's a pretty bad design

Reply:
also, as mentioned above, the access to the debug object (or instance? whatever) needs to be protected from data race.